0005-ordered-quantification.md

```diff
@@ -0,0 +1,644 @@
+
+# Prior art
+[prior-art]: #prior-art
+
+Sequence comprehensions bear a strong resemblance to list comprehensions in Python, which also include similar abilities to 
+bind multiple variables and filter the enumerated values: `[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]`.
+Haskell and Cryptol also feature similar constructs.
+
+Although the syntax `for[each] x in c` is more common, there is precedent for using the symbol `<-` instead in other languages,
+including Scala, Cryptol, and Haskell. Using something other than `in` is recommended to avoid confusion with the `in` boolean operator,
+but I don't have a strong opinion about which symbol or keyword we use.
+
+Most mainstream programming languages include rich libraries for working with collections and iterators,
+and define much the same operations such as "filter" and "flatMap".
+The proposed Dafny features provide similar functionality at a higher level
+of abstraction that will require less effort from Dafny programmers to successfully verify: 
+quantification expressions are used to succinctly declare the behavior of
+an operation that in spirit applies several such operations. See the "Future Possibilities" section for ideas for
+pushing this philosophy even further.
+
+The [JMatch language](https://www.cs.cornell.edu/andru/papers/padl03.pdf), an extension of Java to support
+pattern matching, includes a `foreach` loop remarkably similar to the proposal here: it is parameterized by
+a boolean expression and iterates through all matching bindings. For example, the statement `foreach(a[int i] != 0) n += i;`
+sums the indices of all non-zero elements of the array `a`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+It is very convenient to support ordered quantification over the natural ordering of types, but
+Dafny's choice of built-in ordering concepts is currently somewhat surprising. Historical, Dafny has biased towards
+defining well-founded orderings to support proofs of termination rather than intuitive orderings
+users would expect. For example, `<` on datatypes is defined as "rank comparison", such that `Some(x) < Some(Some(x))`.
+For strings (and more generally sequences), `<` means "proper prefix of", and the well-founded ordering is
+"subsequence of", and neither is the lexicographic comparison users would expect.
+As a somewhat orthogonal improvement, we may want to make breaking changes to the language to align `<` with
+typical programmer expectations and the "natural ordering" concept, as well as defining a separate operator
+for the well-founded orderings (see https://github.com/dafny-lang/dafny/issues/762).
+
+Along similar lines, will users frequently want to customize the ordering of their quantifications without
+creating explicit collection values? To produce the sequence `[4, 3, 2, 1, 0]`, for example, 
+we could consider supporting something like `seq i by > | 0 <= i < 5`.
```

Comment:
> `seq i: int | 0 <= i < 5 :: 4 - i` would not require any change

Yup, that's a nice application of the generic comprehension syntax!

> `seq i: int | 5 > i >= 4 :: i` might seem a bit WTF.

More than just a bit WTF methinks :D That's another case where I don't think we should mess with deeply established semantics of boolean operators. Plus I know @RustanLeino prefers the idiom of always use `<` and `<=` so values are ordered naturally left-to-right in expressions, and I tend to like that too.

> seq i: int | i <- 4 downto 0 :: i

THAT'S not a bad idea, since there's precedent in `for i := 4 downto o { .. }` loops...